assets/app/view/game/discard_trains.rb

```diff
@@ -15,8 +15,9 @@ def render
             verticalAlign: 'top',
           },
         }
-        overflow = @game.active_step.crowded_corps.map do |corporation|
-          trains = corporation.trains.map do |train|
+        step = @game.active_step
+        overflow = step.crowded_corps.map do |corporation|
+          trains = step.trains(corporation).map do |train|
```

Comment:
Get the trains from the step, so shells can be handled without the view needing to know about them.